### PR TITLE
Implement SET PATH statement & add client capabilities header

### PR DESCRIPTION
**Pull request by bowdencm**
Implementation for the SET PATH statement, which sets the path field in the session to later be used for function resolution.

### CODE REVIEW

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java [hunk 1]
```diff
@@ -320,6 +320,13 @@ private static boolean process(QueryRunner queryRunner, String sql, OutputFormat
                 schemaChanged.run();
             }
 
+            //update path if present
+            if (query.getSetPath().isPresent()) {
+                session = ClientSession.builder(session)
+                        .withPath(query.getSetPath().orElse(session.getPath()))
```
Comment (kokosing) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: `orElse()` is redundant, you already checked that the value is present

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestSetPathTask
+{
+    private final TransactionManager transactionManager;
+    private final AccessControl accessControl;
+    private final MetadataManager metadata;
+
+    private final ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("stage-executor-%s"));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: it looks like you do not close the `executor`

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.transaction.TransactionManager.createTestTransactionManager;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Collections.emptyList;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestSetPathTask
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: Tests would be much easier if you would provide `SHOW PATH`

**Comment by findepi** [comment]
I am not entirely a fan of how session state change is being managed, although I know this is in line with other session-changing statements. We have been here with Roles and it turned out to be problematic, because you could have a client that works for everything except the new statement, which modifies session state. Statement executes nicely, but un-updated client will ignore the additional headers. Here, it is possible that a client  executes `set path <...>; query1; query2; query3; query4;` and `query4` fails because `set path` was ignored. 

I think we should have a different strategy for statements, which sole purpose is to change session state -- like `SET PATH`, `USE` or `SET ROLE` (https://github.com/prestodb/presto/pull/10904). We should return session state delta in a way that un-updated client will not ignore, it would fail instead. Perhaps, this could still be a header, like `X-Presto-Session-State-Change: path: <new-path>`. Then, each client impl needs to know only this one header `X-Presto-Session-State-Change` to be able to bail out on unsupported session state change.

@electrum @kokosing thoughts?

**Comment by kokosing** [comment]
@findepi Great idea.That way it will be explicit if client is not supported.

> X-Presto-Session-State-Change: path: <new-path>.

Regarding the above state change or any other, I wonder how that influence the client itself. Naive solution would to make client just to save anything stored in that property and send it back to the server.

**Comment by findepi** [comment]
> Naive solution would to make client just to save anything stored in that property and send it back to the server.

True, we could follow up with that. For `USE` it won't "just work" -- CLI shows that in prompt so needs to understand that. For everything else -- I think more or less opaque approach could do (maybe not fully opaque, if we want to send only deltas back to the client, as we do now).

**Comment by electrum** [comment]
@findepi I completely agree. This is something we want to solve with the [v2 client protocol](https://github.com/prestodb/presto/wiki/Presto-v2-client-protocol). In the current draft, we have an explicit list of actions, and clients should fail if they do not understand any of the actions. We also discussed if the session state should be a more generic "cookie" style mechanism, which could allow additional features like this to work automatically with older clients. That leads to the question of which things are special or well known and have first class support. For example, if clients are not aware of things like catalog and schema, they would need to execute a `USE` statement for every connection.

**Comment by electrum** [comment]
For the v1 protocol, I think we should continue on the same route of using first class headers, in order to keep the protocol consistent. However, we can prevent silent breakage if we require clients to send a capability list:
```
X-Presto-Client-Capabilities: path
```
Have the clients send that header and reject `SET PATH` if the capabilities list does not contain `path`.

**Comment by findepi** [comment]
Since v1 will still be around for a while, l like the capabilities approach that (in practice) closes the silent breakage problem.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 4]
```diff
@@ -550,6 +552,16 @@ public void setSetSchema(String schema)
         setSchema.set(requireNonNull(schema, "schema is null"));
     }
 
+    public void setSetPath(String path)
+    {
+        setPath.set(requireNonNull(path, "path is null"));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: nit: I know that in other parts of this class we put the requireNonNull calls inside another method call, but let's not do it here for the sake of slightly better readability:

```
requireNonNull(path, "path is null");
setPath.set(path);
```

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java [hunk 5]
```diff
@@ -32,8 +43,33 @@ public String getName()
...
+            List<Expression> parameters)
     {
-        throw new UnsupportedOperationException("SET PATH not implemented yet");
+        Session session = stateMachine.getSession();
+
+        //convert to IR before setting HTTP headers - ensures that the representations of all path objects outside the parser remain consistent
+        SqlPath sqlPath = new SqlPath(Optional.of(statement.getPathSpecification().toString()));
+        List<SqlPathElement> path = sqlPath.getParsedPath();
+
+        for (SqlPathElement element : path) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: `path` is not used anywhere else, so you can do 

```
for (SqlPathElement element : sqlPath.getParsedPath())
```

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java [hunk 6]
```diff
@@ -140,6 +140,10 @@ public Session createSession(QueryId queryId, SessionContext context, Optional<S
                 .setTraceToken(context.getTraceToken())
                 .setResourceEstimates(context.getResourceEstimates());
 
+        if (context.getPath() != null) {
+            sessionBuilder.setPath(new SqlPath(Optional.of(context.getPath())));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: According to `SessionContext` interface `getPath` is `Nullable`, so we should call `Optional.ofNullable(context.getPath())`.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                Optional.empty(), ImmutableList.of(
+                        new PathElement(Optional.empty(), new Identifier("foo"))));
+
+        QueryStateMachine stateMachine = getStateMachine("SET PATH foo");
+        executeSetPathTask(pathSpecification, stateMachine);
+
+        assertEquals(stateMachine.getSetPath(), "foo");
+    }
+
+    @Test (expectedExceptions = PrestoException.class)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: Please add `expectedExceptionsMessageRegExp`.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    public void testSetPathInvalidCatalog()
+    {
+        PathSpecification invalidPathSpecification = new PathSpecification(
+                Optional.empty(), ImmutableList.of(
+                        new PathElement(Optional.of(new Identifier("invalidCatalog")), new Identifier("thisDoesNotMatter"))));
+
+        QueryStateMachine stateMachine = getStateMachine("SET PATH invalidCatalog.thisDoesNotMatter");
+        executeSetPathTask(invalidPathSpecification, stateMachine);
+        //error should have been thrown due to nonexistent catalog name
+        fail();
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: This `fail` call and the comment above it are not needed, because if this test doesn't throw a `PrestoException` it will be marked as failed by testng.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private QueryStateMachine getStateMachine(String query)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: `getStateMachine ` -> `createQueryStateMachine`.

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java (hunk 6 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: Even after the check to make sure that it's not null?

Comment (nezihyigitbasi) [reply presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: Oops I missed that, you are right.

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java [hunk 10]
```diff
@@ -320,6 +320,13 @@ private static boolean process(QueryRunner queryRunner, String sql, OutputFormat
                 schemaChanged.run();
             }
 
+            //update path if present
+            if (query.getSetPath().isPresent()) {
```
Comment (kokosing) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: Would not be better to create `ClientSession.Builder` in line 312 and just update path here instead of creating new `session`, then in line 360 you could just build a session.

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java [hunk 11]
```diff
@@ -32,8 +43,32 @@ public String getName()
...
 
     @Override
-    public ListenableFuture<?> execute(SetPath statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    public ListenableFuture<?> execute(
+            SetPath statement,
+            TransactionManager transactionManager,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: please undo this change if there is nothing actually changed but formatting. It is nice to have formatting extracted as separate commits.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResource.java [hunk 12]
```diff
@@ -192,9 +193,10 @@ private static Response toResponse(Query query, QueryResults queryResults)
     {
         ResponseBuilder response = Response.ok(queryResults);
 
-        // add set catalog and schema
+        // add set catalog, schema, and path
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/StatementResource.java]: I would just remove this comment. It is trivial and code is self explaining here.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                new SessionPropertyManager(),
+                new SchemaPropertyManager(),
+                new TablePropertyManager(),
+                transactionManager);
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        executor.shutdownNow();
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: set it to `null` afterwards like we do in other tests

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+                Optional.empty(), ImmutableList.of(
+                        new PathElement(Optional.empty(), new Identifier("foo"))));
+
+        QueryStateMachine stateMachine = createQueryStateMachine("SET PATH foo");
+        executeSetPathTask(pathSpecification, stateMachine);
+
+        assertEquals(stateMachine.getSetPath(), "foo");
+    }
+
+    @Test (expectedExceptions = PrestoException.class, expectedExceptionsMessageRegExp = "Catalog does not exist: .*")
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: use `assertThatThrownBy ` from http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html#exception-assertion

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java [hunk 15]
```diff
@@ -91,6 +92,9 @@
             "/" +
             firstNonNull(StatementClientV1.class.getPackage().getImplementationVersion(), "unknown");
 
+    //add new elements as a comma-separated list, eg. "path, foo, bar"
+    private static final String CLIENT_CAPABILITIES = "path";
```
Comment (kokosing) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java]: why not to keep it as a plain list (or even set) instead of string?

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java [hunk 16]
```diff
@@ -53,6 +54,10 @@ public String getName()
     {
         Session session = stateMachine.getSession();
 
+        if (!session.getClientCapabilities().contains("path")) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: should not this be something checked at the lower level, closer to the communication? When state machine have path set then we could check if client is capable to handle the path before sending him a response. What do you think?

Here it is simple, but in case of roles (https://github.com/prestodb/presto/pull/10904) we would need to check in many places as there are many role related statements. Would we need to update all of these statements execution code? I would prefer to change a single place instead.

Also here you are mixing domains: SQL execution with client communication.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java (hunk 13 above)
Comment (bowdencm) [reply presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: I have it declared as final (which is in line with other tests as well), so it can't be re-assigned to null.

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java [hunk 17]
```diff
@@ -320,6 +320,13 @@ private static boolean process(QueryRunner queryRunner, String sql, OutputFormat
                 schemaChanged.run();
             }
 
+            //update path if present
```
Comment (electrum) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: Nit: use space after `//`. Same for all comments.

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java (hunk 16 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: We don't really have a way to do so - we want to fail the query itself, so if we put the check later (eg in StatementResource), then the query will have completed and the error won't behave as we wanted.

File: presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java (hunk 14 above)
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSetPathTask.java]: Nit: no space between the annotation and the opening `(`

(this will probably be caught by Checkstyle)

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java (hunk 15 above)
Comment (bowdencm) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java]: I think it's simplest to use a string when the header is going to be set to a string anyway.

Comment (nezihyigitbasi) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java]: I also think that we shouldn't maintain the capabilities as a string. How about we have an enum, say `ClientCapabilities`, you can then create this static final String with `Joiner.on(",").join(ClientCapabilities.values())`. You can use that enum in other parts of the code (e.g., you hardcode "path" in `SetPathTask`).

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 18]
```diff
@@ -291,6 +300,12 @@ public boolean supportClientTransaction()
         return ImmutableSet.copyOf(splitter.split(nullToEmpty(servletRequest.getHeader(PRESTO_CLIENT_TAGS))));
     }
 
+    private Set<String> parseClientCapabilities(HttpServletRequest servletRequest)
+    {
+        Splitter splitter = Splitter.on(',').trimResults().omitEmptyStrings();
+        return ImmutableSet.copyOf(splitter.split(nullToEmpty(servletRequest.getHeader(PRESTO_CLIENT_CAPABILITIES))));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: Do we want to support clients putting arbitrary strings in this header? Or do we just want to fail loudly?

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java [hunk 19]
```diff
@@ -311,52 +311,46 @@ private static boolean process(QueryRunner queryRunner, String sql, OutputFormat
...
+            ClientSession.Builder builder = ClientSession.builder(session);
+
+            if (query.getStartedTransactionId() != null) {
+                builder = builder.withTransactionId(query.getStartedTransactionId());
+            }
+
             // update catalog and schema if present
             if (query.getSetCatalog().isPresent() || query.getSetSchema().isPresent()) {
-                session = ClientSession.builder(session)
+                builder = builder
```
Comment (nezihyigitbasi) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]:
```
builder = builder.withCatalog(query.getSetCatalog().orElse(session.getCatalog()))
        .withSchema(query.getSetSchema().orElse(session.getSchema()));
```

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java (hunk 18 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: "arbitrary" may actually mean "forward-compatible", newer client talking to older server
